lib/frodata/config.rb

```diff
@@ -0,0 +1,146 @@
+# frozen_string_literal: true
+
+require 'logger'
+
+module FrOData
+  class << self
+    attr_writer :log
+
+    # Returns the current Configuration
+    #
+    # Example
+    #
+    #    FrOData.configuration.username = "username"
+    #    FrOData.configuration.password = "password"
+    def configuration
+      @configuration ||= Configuration.new
+    end
+
+    # Yields the Configuration
+    #
+    # Example
+    #
+    #    FrOData.configure do |config|
+    #      config.username = "username"
+    #      config.password = "password"
+    #    end
+    def configure
+      yield configuration
+    end
+
+    def log?
+      @log ||= false
+    end
+
+    def log(message)
+      return unless FrOData.log?
+      configuration.logger.send(configuration.log_level, message)
+    end
+  end
+
+  class Configuration
+    class Option
+      attr_reader :configuration, :name, :options
+
+      def self.define(*args)
+        new(*args).define
+      end
+
+      def initialize(configuration, name, options = {})
+        @configuration = configuration
+        @name = name
+        @options = options
+        @default = options.fetch(:default, nil)
+      end
+
+      def define
+        write_attribute
+        define_method if default_provided?
+        self
+      end
+
+      private
+
+      attr_reader :default
+      alias default_provided? default
+
+      def write_attribute
+        configuration.send :attr_accessor, name
+      end
+
+      def define_method
+        our_default = default
+        our_name    = name
+        configuration.send :define_method, our_name do
+          instance_variable_get(:"@#{our_name}") ||
+            instance_variable_set(
+              :"@#{our_name}",
+              our_default.respond_to?(:call) ? our_default.call : our_default
+            )
+        end
+      end
+    end
+
+    class << self
+      attr_accessor :options
+
+      def option(*args)
+        option = Option.define(self, *args)
+        (self.options ||= []) << option.name
+      end
+    end
+
+    option :api_version, default: lambda { ENV['DYNAMICS_API_VERSION'] || '9.1' }
```

Comment:
Technically if this is a more generic OData library env variables for Dynamics shouldn't really be here

Reply:
Maybe just prefixing with ODATA instead for some of these. Also have it default to 9.1 is just way to specific. I would just default to no version, I'm not even sure if version should be a parameter and it should just be part of the uri, something else can build that in. But that's my 2 cents.

Reply:
yea I was trying to map Dynamics. Will figure out how extrac the path and that version out of the gem